src/chain_sync/chain_follower.rs

```diff
@@ -0,0 +1,969 @@
+
+                    sync_states_guard.truncate(std::num::NonZeroUsize::new(1).unwrap());
+                    let first = sync_states_guard.first_mut();
+                    first.set_epoch(heaviest.epoch());
+                    first.set_target(Some(
+                        state_machine
+                            .lock()
+                            .heaviest_tipset()
+                            .unwrap_or(heaviest.clone()),
+                    ));
+                    let seconds_per_epoch = state_manager.chain_config().block_delay_secs;
+                    let time_diff =
+                        (Utc::now().timestamp() as u64).saturating_sub(heaviest.min_timestamp());
+                    if time_diff < seconds_per_epoch as u64 * 2 {
+                        first.set_stage(SyncStage::Complete);
+                    } else {
+                        first.set_stage(SyncStage::Messages);
+                    }
+                    sync_states_guard.extend(states);
+                }
+
+                for task in task_vec {
+                    // insert task into tasks. If task is already in tasks, skip. If it is not, spawn it.
+                    let new = tasks_set.insert(task.clone());
+                    if new {
+                        let tasks_clone = tasks.clone();
+                        let action = task.clone().execute(
+                            network.clone(),
+                            state_manager.clone(),
+                            bad_block_cache.clone(),
+                            stateless_mode,
+                        );
+                        tokio::spawn({
+                            let state_machine = state_machine.clone();
+                            let state_changed = state_changed.clone();
+                            async move {
+                                if let Some(event) = action.await {
+                                    state_machine.lock().update(event);
+                                    state_changed.notify_one();
+                                }
+                                tasks_clone.lock().remove(&task);
+                            }
+                        });
+                    }
+                }
+            }
+        }
+    });
+
+    // Periodically report progress if there are any tipsets left to be fetched.
+    // Once we're in steady-state (i.e. caught up to HEAD) and there are no
+    // active forks, this will not report anything.
+    set.spawn({
+        let state_manager = state_manager.clone();
+        let state_machine = state_machine.clone();
+        async move {
+            loop {
+                tokio::time::sleep(tokio::time::Duration::from_secs(10)).await;
+                let (tasks_set, _) = state_machine.lock().tasks();
+                let heaviest_epoch = state_manager.chain_store().heaviest_tipset().epoch();
+                let fork_cutoff = heaviest_epoch
```

Comment:
A day seems arbitrary. Shouldn't we use `CHAIN_FINALITY`? It's even smaller (but still considered conservative).

Reply:
We previously used 1 day. Let's keep it and lower it in another PR. Don't want to change too many things at once.